// css-literal-loader/lib/traverse.js
"use strict";

exports.__esModule = true;
exports.default = traverse;

var _babylon = require("babylon");

var _codeFrame = require("@babel/code-frame");

var _traverse2 = _interopRequireWildcard(require("@babel/traverse"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function parseSource(src) {
  return (0, _babylon.parse)(src, {
    allowImportExportEverywhere: true,
    allowReturnOutsideFunction: true,
    allowSuperOutsideMethod: true,
    sourceType: 'unambigious',
    sourceFilename: true,
    plugins: ['jsx', 'flow', 'doExpressions', 'objectRestSpread', 'decorators', 'classProperties', 'classPrivateProperties', 'classPrivateMethods', 'exportDefaultFrom', 'exportNamespaceFrom', 'asyncGenerators', 'functionBind', 'functionSent', 'dynamicImport', 'numericSeparator', 'optionalChaining', 'importMeta', 'bigInt', 'optionalCatchBinding', 'throwExpressions', 'pipelineOperator', 'nullishCoalescingOperator']
  });
}

function traverse(source, visitors, opts) {
  var ast = parseSource(source); // https://github.com/babel/babel/issues/4640 (closed but reverted)

  var hub = new _traverse2.Hub({
    buildCodeFrameError: function buildCodeFrameError(node, message) {
      var msg = message; // eslint-disable-next-line no-underscore-dangle

      var loc = node && (node.loc || node._loc);

      if (loc) {
        msg = msg + " (" + loc.start.line + ":" + loc.start.column + ")\n";
        msg += (0, _codeFrame.codeFrameColumns)(source, loc);
      }

      return new Error(msg);
    }
  });

  var path = _traverse2.NodePath.get({
    hub: hub,
    parentPath: null,
    parent: ast,
    container: ast,
    key: 'program'
  });

  return (0, _traverse2.default)(ast, visitors, path.setContext().scope, {
    opts: opts
  });
}
